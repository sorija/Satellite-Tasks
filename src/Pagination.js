import React from "react";
import { connect } from "react-redux";
import { setCurrentPage } from "./actions/paginationActions.js";

class Pagination extends React.Component {
  render() {
    const { currentPage, setCurrentPage } = this.props;
    // number of pages is equal to number of tasks divided by limit of tasks per page and rounded up
    const length = this.props.requests.length / 5;
    const pagination = Array.from(
      { length: Math.ceil(length) },
      (_, i) => i + 1
    );
    return (
      <React.Fragment>
        <button
          type="button"
          disabled={currentPage === 1}
          onClick={() => setCurrentPage(1)}
        >
          &laquo;
        </button>
        <button
          type="button"
          disabled={currentPage === 1}
          onClick={() => setCurrentPage(currentPage - 1)}
        >
          &lt;
        </button>
        {pagination.map(page => (
          <button
            key={page}
            type="button"
            disabled={currentPage === page}
            onClick={() => setCurrentPage(page)}
          >
            {page}
          </button>
        ))}
        <button
          type="button"
          disabled={currentPage === pagination.length}
          onClick={() => setCurrentPage(currentPage + 1)}
        >
          &gt;
        </button>
        <button
          type="button"
          disabled={currentPage === pagination.length}
          onClick={() => setCurrentPage(pagination.length)}
        >
          &raquo;
        </button>
      </React.Fragment>
    );
  }
}

const mapStateToProps = state => {
  return {
    requests: state.requests,
    currentPage: state.currentPage
  };
};

const mapDispatchToProps = dispatch => ({
  setCurrentPage: page => dispatch(setCurrentPage(page))
});
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Pagination);
