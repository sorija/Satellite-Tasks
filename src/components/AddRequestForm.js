import React from "react";

export class AddRequestForm extends React.Component {
  // using local state to track changes without modifying data before submitting
  constructor(props) {
    super(props);
    this.state = {
      satName: "",
      longitude: "",
      latitude: ""
    };
  }
  validateNumImput = num => {
    return !num || num.match(/^\d+(\.\d*)?$/);
  };
  onNameChange = e => {
    // store the e.target.value in a variable or event will be nullified before the .setState() fires
    const satName = e.target.value;
    this.setState(() => ({ satName }));
  };
  onLongitudeChange = e => {
    const longitude = e.target.value;
    if (this.validateNumImput(longitude)) {
      this.setState(() => ({ longitude }));
    }
  };
  onLatitudeChange = e => {
    const latitude = e.target.value;
    if (this.validateNumImput(latitude)) {
      this.setState(() => ({ latitude }));
    }
  };
  onSubmit = e => {
    e.preventDefault();
    this.props.onSubmit({
      satName: this.state.satName,
      longitude: this.state.longitude,
      latitude: this.state.latitude
    });
  };
  render() {
    const { satName, longitude, latitude } = this.state;
    return (
      <form onSubmit={this.onSubmit}>
        <input
          type="text"
          placeholder="Satellite's Name"
          value={satName}
          onChange={this.onNameChange}
          autoFocus
        />
        <input
          type="text"
          placeholder="Longitude"
          value={longitude}
          onChange={this.onLongitudeChange}
        />
        <input
          type="text"
          placeholder="Latitude"
          value={latitude}
          onChange={this.onLatitudeChange}
        />
        <input
          disabled={!satName || !longitude || !latitude}
          type="submit"
          value="Submit"
        />
      </form>
    );
  }
}

export default AddRequestForm;
