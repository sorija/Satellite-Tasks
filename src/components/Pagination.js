import React from "react";
import { connect } from "react-redux";
import { setCurrentPage } from "../actions/pagination.js";

export class Pagination extends React.Component {
  render() {
    const { currentPage, setCurrentPage } = this.props;
    // number of pages is equal to number of tasks divided by limit of tasks per page and rounded up
    const length = this.props.requests.length / 5;
    const pagination = Array.from(
      { length: Math.ceil(length) },
      (_, i) => i + 1
    );
    return (
      <React.Fragment>
        <button
          id="firstPageBtn"
          type="button"
          disabled={currentPage === 1}
          onClick={() => setCurrentPage(1)}
        >
          &laquo;
        </button>
        <button
          id="prevPageBtn"
          type="button"
          disabled={currentPage === 1}
          onClick={() => setCurrentPage(currentPage - 1)}
        >
          &lt;
        </button>
        {pagination.map(page => (
          <button
            className="pageBtn"
            key={page}
            type="button"
            disabled={currentPage === page}
            onClick={() => setCurrentPage(page)}
          >
            {page}
          </button>
        ))}
        <button
          id="nextPageBtn"
          type="button"
          disabled={currentPage === pagination.length}
          onClick={() => setCurrentPage(currentPage + 1)}
        >
          &gt;
        </button>
        <button
          id="lastPageBtn"
          type="button"
          disabled={currentPage === pagination.length}
          onClick={() => setCurrentPage(pagination.length)}
        >
          &raquo;
        </button>
      </React.Fragment>
    );
  }
}

const mapStateToProps = state => {
  return {
    requests: state.requests,
    currentPage: state.currentPage
  };
};

const mapDispatchToProps = dispatch => ({
  setCurrentPage: page => dispatch(setCurrentPage(page))
});
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Pagination);
