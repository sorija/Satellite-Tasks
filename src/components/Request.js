import React from "react";

const Request = ({ id, satName, longitude, latitude, done, onChange }) => (
  <div className="request-container">
    <div className="name-and-id">
      <p className="request-id data">
        ID: <span className="data-span">{id}</span>
      </p>
      <p className="sat-name">{satName}</p>
    </div>
    <div className="coordinates">
      <p className="lat data">
        Latitude: <span className="data-span">{latitude}</span>
      </p>
      <p className="lon data">
        Longitude: <span className="data-span">{longitude}</span>
      </p>
    </div>
    <input
      className="request-status data"
      type="checkbox"
      value={id}
      checked={done}
      onChange={onChange}
    />
  </div>
);

export default Request;
