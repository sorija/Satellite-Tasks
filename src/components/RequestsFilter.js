import React from "react";
import { connect } from "react-redux";
import { setTextFilter, setStatusFilter } from "../actions/filters.js";

class RequestsFilter extends React.Component {
  onTextChange = e => {
    this.props.setTextFilter(e.target.value);
  };
  onStatusChange = e => {
    this.props.setStatusFilter(e.target.value);
  };
  render() {
    return (
      <div>
        <input
          type="text"
          placeholder="Search By Satellite..."
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <label htmlFor="status">Show: </label>
        <select id="status" onChange={this.onStatusChange}>
          <option value="all">All</option>
          <option value="new">New</option>
          <option value="done">Done</option>
        </select>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  setStatusFilter: status => dispatch(setStatusFilter(status))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(RequestsFilter);
