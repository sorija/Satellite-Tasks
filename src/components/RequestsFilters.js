import React from "react";
import { connect } from "react-redux";
import { setTextFilter, setStatusFilter } from "../actions/filters.js";

export class RequestsFilters extends React.Component {
  onTextChange = e => {
    this.props.setTextFilter(e.target.value);
  };
  onStatusChange = e => {
    this.props.setStatusFilter(e.target.value);
  };
  render() {
    return (
      <div className="filters-container">
        <div className="select-container">
          <label htmlFor="status">Show: </label>
          <select id="status" onChange={this.onStatusChange}>
            <option value="all">All</option>
            <option value="new">New</option>
            <option value="done">Done</option>
          </select>
        </div>
        <input
          id="text-filter"
          type="text"
          placeholder="Search By Satellite..."
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  setStatusFilter: status => dispatch(setStatusFilter(status))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(RequestsFilters);
