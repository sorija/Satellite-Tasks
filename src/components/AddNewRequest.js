import React from "react";
import { connect } from "react-redux";
import AddRequestForm from "./AddRequestForm.js";
import { addRequest } from "../actions/requests.js";

export class AddNewRequest extends React.Component {
  constructor(props) {
    super(props);
    this.state = { addNew: false };
  }
  handleAddClick = () => {
    this.setState({ addNew: !this.state.addNew });
  };
  onSubmit = ({ satName, longitude, latitude }) => {
    this.props.addRequest(satName, parseFloat(longitude), parseFloat(latitude));
    this.setState({ addNew: false });
  };
  render() {
    let addRequestForm;
    if (this.state.addNew) {
      addRequestForm = <AddRequestForm onSubmit={this.onSubmit} />;
    }
    return (
      <div className="add-request-container">
        <button className="add-request__btn" onClick={this.handleAddClick}>
          NEW REQUEST
        </button>
        {addRequestForm}
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addRequest: (satName, longitude, latitude) =>
    dispatch(addRequest(satName, longitude, latitude))
});

export default connect(
  null,
  mapDispatchToProps
)(AddNewRequest);
