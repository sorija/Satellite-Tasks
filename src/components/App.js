import React, { Component } from "react";
import { Provider } from "react-redux";
import configureStore from "../configureStore.js";
import RequestsList from "./RequestsList.js";
import AddNewRequest from "./AddNewRequest.js";
import RequestsFilters from "./RequestsFilters.js";
import "../styles/App.css";

const store = configureStore();

class App extends Component {
  render() {
    return (
      <Provider store={store}>
        <div className="App">
          <header className="App-header">
            <p>Hello Space!</p>
          </header>
          <RequestsFilters />
          <AddNewRequest />
          <RequestsList />
        </div>
      </Provider>
    );
  }
}

export default App;
