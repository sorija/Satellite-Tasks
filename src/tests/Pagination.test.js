import React from "react";
import { shallow } from "enzyme";
import { Pagination } from "../components/Pagination";

let wrapper, mockSetCurrentPage;
beforeEach(() => {
  mockSetCurrentPage = jest.fn();
  wrapper = shallow(
    <Pagination
      requests={[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]}
      setCurrentPage={mockSetCurrentPage}
    />
  );
});

test("renders Pagination component", () => {
  expect(wrapper).toMatchSnapshot();
});

test("renders correct number of page-specific buttons", () => {
  let buttons = wrapper.find("button");
  expect(buttons).toHaveLength(7); //3 for pages + 4 for navigation
});
