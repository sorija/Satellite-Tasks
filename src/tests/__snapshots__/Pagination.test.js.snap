// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Pagination component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Pagination
    requests={
      Array [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
      ]
    }
    setCurrentPage={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          disabled={false}
          id="firstPageBtn"
          onClick={[Function]}
          type="button"
        >
          «
        </button>,
        <button
          disabled={false}
          id="prevPageBtn"
          onClick={[Function]}
          type="button"
        >
          &lt;
        </button>,
        Array [
          <button
            className="pageBtn"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            1
          </button>,
          <button
            className="pageBtn"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            2
          </button>,
          <button
            className="pageBtn"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            3
          </button>,
        ],
        <button
          disabled={false}
          id="nextPageBtn"
          onClick={[Function]}
          type="button"
        >
          &gt;
        </button>,
        <button
          disabled={false}
          id="lastPageBtn"
          onClick={[Function]}
          type="button"
        >
          »
        </button>,
      ],
      "className": "pages-nav",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "«",
          "disabled": false,
          "id": "firstPageBtn",
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": "«",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "<",
          "disabled": false,
          "id": "prevPageBtn",
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": "<",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "host",
        "props": Object {
          "children": 1,
          "className": "pageBtn",
          "disabled": false,
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": 1,
        "type": "button",
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "host",
        "props": Object {
          "children": 2,
          "className": "pageBtn",
          "disabled": false,
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": 2,
        "type": "button",
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "host",
        "props": Object {
          "children": 3,
          "className": "pageBtn",
          "disabled": false,
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": 3,
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": ">",
          "disabled": false,
          "id": "nextPageBtn",
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": ">",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "»",
          "disabled": false,
          "id": "lastPageBtn",
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": "»",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            disabled={false}
            id="firstPageBtn"
            onClick={[Function]}
            type="button"
          >
            «
          </button>,
          <button
            disabled={false}
            id="prevPageBtn"
            onClick={[Function]}
            type="button"
          >
            &lt;
          </button>,
          Array [
            <button
              className="pageBtn"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              1
            </button>,
            <button
              className="pageBtn"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              2
            </button>,
            <button
              className="pageBtn"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              3
            </button>,
          ],
          <button
            disabled={false}
            id="nextPageBtn"
            onClick={[Function]}
            type="button"
          >
            &gt;
          </button>,
          <button
            disabled={false}
            id="lastPageBtn"
            onClick={[Function]}
            type="button"
          >
            »
          </button>,
        ],
        "className": "pages-nav",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "«",
            "disabled": false,
            "id": "firstPageBtn",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": "«",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "<",
            "disabled": false,
            "id": "prevPageBtn",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": "<",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "host",
          "props": Object {
            "children": 1,
            "className": "pageBtn",
            "disabled": false,
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": 1,
          "type": "button",
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "host",
          "props": Object {
            "children": 2,
            "className": "pageBtn",
            "disabled": false,
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": 2,
          "type": "button",
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "host",
          "props": Object {
            "children": 3,
            "className": "pageBtn",
            "disabled": false,
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": 3,
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": ">",
            "disabled": false,
            "id": "nextPageBtn",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": ">",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "»",
            "disabled": false,
            "id": "lastPageBtn",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": "»",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
